### ISSUE TITLE
Partition pushdown through window functions

### ISSUE THREAD

**Issue by grantatspothero** [prestosql/presto#1272]
Some partition filters are pushed down through window functions while some are not. 

Partition filters are pushed down through window functions if the `partition_by` clause of the window function is the partition key (along with some other caveats). 

Example: test_table partitioned by dateid and partition pushdown works
```
explain
select * from (
  select *, row_number() over(partition by dateid order by another_id desc) as rn
  from test_table
)
where dateid >= 20190501 and rn = 1 
```

Partition filters are *not* pushed down if ordering by a partition key with an explicit window frame.
Example: test_table partition by dateid and partition pushdown does not work. Presto only needs to look at the date partitions from [20190429:current_date] but instead will do a full table scan.  
```
explain
select * from (
  select *, sum(some_value) over(order by dateid desc rows between 2 preceding and current ROW) as rn
  from hive.segment.tracks 
)
where dateid >= 20190501
```

Sorry for the contrived examples above, I simplified some actual queries to make the problem clearer. My actual use case is trying to create a rolling count of daily active users/weekly active users/monthly active users as follows:
```
explain
with test as (
  select
    dateid,
    cardinality(hll) as approx_DAU,
    cardinality(
        merge(hll) over(
          order by dateid asc ROWS BETWEEN 6 PRECEDING AND current row
        )
    ) as approx_WAU,
    cardinality(
        merge(hll) over(
          order by dateid asc ROWS BETWEEN 29 PRECEDING AND current row
        )
    ) as approx_MAU
    from (
      select
        tracks.dateid,
        approx_set(user_id) as hll
      from hive.clickstream_data -- this table is partitioned by dateid as an int 
      group by 1
    )
)
select * 
from test
where dateid >= 20190501 
``` 
And this will do a full table scan when it only needs to look at 29 date partitions before 20190501.

**Comment by sopel39** [comment]
Thanks for reporting. Can you reproduce the issue using TPCH connector (e.g filter is not being pushed down to table scan)?

Can you also attach output of `EXPLAIN (TYPE DISTRIBUTED)` for your query?

**Comment by sopel39** [comment]
Hi @grantatspothero 

I think this is pretty sophisticated optimization. The idea is really interesting though.
Is this a practical issue for you? As a workaround you could add explicit filter before window aggregation is applied.

**Comment by grantatspothero** [comment]
@sopel39 you're correct that if you can modify the query, a workaround is manually pushing down the filter

The practical problem is counting daily/weekly/monthly active users using HLL (see the query above).

This query is generated by a BI dashboard tool (Looker) and getting that tool to add in extra filters is difficult.

**Comment by sopel39** [comment]
The sophisticated part in this issue is that Presto would need to pushdown some kind of ranked range to the connector. This could be done, but it's pretty complex.
